[Operator Versioning][Edge] Implement register function for upgrader

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67731
* #67728
* __->__ #67730



This pr implement the register function for upgrader so it can be used at loading stage

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518/)

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518)

Commit:
[Operator Versioning][Edge] Implement register function for upgrader

This pr implement the register function for upgrader so it can be used at loading stage

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518/)

[ghstack-poisoned]

zhxchen17 commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -303,6 +304,14 @@ void BytecodeDeserializer::parseFunctionSchema(
 }
 
+void BytecodeDeserializer::init_upgrader(mobile::Function* function) {
+  for (auto const& [upgrader_name, upgrader_bytecode] : kUpgraderBytecode) {
```
PyTorch OSS uses C++14, so unfortunately we need to avoid structured binding.

cccclai replied:
Will be using kUpgradeFunction, so it's not needed any more

zhxchen17 commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -303,6 +304,14 @@ void BytecodeDeserializer::parseFunctionSchema(
 
+void BytecodeDeserializer::init_upgrader(mobile::Function* function) {
+  for (auto const& [upgrader_name, upgrader_bytecode] : kUpgraderBytecode) {
+    std::unique_ptr<mobile::Function> func = mobile::Function::get(
```
`func` will be released immediately at the end of the loop body, consider storing these functions in a static local container.

cccclai replied:
Yeah resolved by using static

zhxchen17 reviewed (commented):
overall looks exactly right to me, thanks for the great work :) some small things.

Commit:
Update on "[Operator Versioning][Edge] Implement register function for upgrader"


This pr implement the register function for upgrader so it can be used at loading stage

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518/)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Implement register function for upgrader"


This pr implement the register function for upgrader so it can be used at loading stage

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518/)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Implement register function for upgrader"




This pr implement the register function for upgrader so it can be used at loading stage

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518/)

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -197,6 +200,34 @@ const std::vector<int64_t>& Function::getExceptionDebugHandles() const {
+    const std::vector<c10::TypePtr> types,
+    const google::int64 register_size) {
+  static Function func = Function(c10::QualifiedName(qualified_name));
+  for (auto const& inst : instructions) {
```
I think this line an below would still be executed even with static in line 210.

cccclai replied:
Yeah that's right. I was thinking that the instruction and etc will still be initialized here?

iseeyuan replied:
If it's initialized here, it should be initialized only once. Although func is static, it's instructions, constants, types, etc, are appended each time this function is called. If there are multiple upgraders, this func would have all instructions from different upgraders. A simple example: 
```C++
#include <iostream>
#include <string>
#include <vector>

std::string& GetAString(int i) {
    static std::string s;
    s += std::to_string(i);
    return s;
}


int main()
{
    std::vector<std::string> vs;
    for (int i = 0; i < 3; ++i) {
        vs.emplace_back(GetAString(i));
    }
    
    for (const auto& s : vs) {
        std::cout << s << " ";
    }

    return 0;
}
```
output:
```
0 01 012 
```

So the entire upgrader table needs to be static, like @zhxchen17 suggested in another comment. If the same upgrader appears the second time, we check if the name is already in the registry, to avoid appending the elements again.

cccclai replied:
Hmm the entire upgrade table is static, in `upgrader_mobile.h`. It's like following and this function will only be called once.

```
static std::vector<ByteCodeFunctionWithOperator> kUpgraderByteCode(
    {ByteCodeFunctionWithOperator({
        mobile::Function::registerFunc(
            "div_Tensor_0_3",
            std::vector<Instruction>({
                Instruction{OpCode::STOREN, 1, 2},
                Instruction{OpCode::LOAD, 1, 0},
                Instruction{OpCode::OP, 0, 0},
                ****
                Instruction{OpCode::RET, 0, 0},
            }), // instructions_
            std::vector<c10::IValue>({
                c10::IValue("trunc"),
                c10::IValue(true),
            }), // constants
            std::vector<c10::TypePtr>(), // types
            4 // register_size_
            ),
        std::vector<OperatorString>({
            OperatorString({"aten::is_floating_point", "", 1}),
            OperatorString({"aten::div", "Tensor", 2}),
            OperatorString({"aten::div", "Tensor_mode", 3}),
        }), // op_names

    })});
```
Do I miss anything?

cccclai replied:
Address in latest commit.

Commit:
Update on "[Operator Versioning][Edge] Implement register function for upgrader"




This pr implement the register function for upgrader so it can be used at loading stage

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518/)

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -214,6 +218,26 @@ c10::optional<std::function<void(Stack&)>> makeOperatorFunction(
+    const std::vector<c10::IValue> constants,
+    const std::vector<c10::TypePtr> types,
+    const google::int64 register_size) {
+  static Function func = Function(c10::QualifiedName(qualified_name));
```
`static` here may still cause potential issue of adding instructions of different upgraders to the same function.

cccclai replied:
My changes are not updated...sorry about that and I push a new commit.

Commit:
Update on "[Operator Versioning][Edge] Implement register function for upgrader"




This pr implement the register function for upgrader so it can be used at loading stage

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518/)

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Implement register function for upgrader"




This pr implement the register function for upgrader so it can be used at loading stage

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518/)

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Implement register function for upgrader"




This pr implement the register function for upgrader so it can be used at loading stage

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518/)

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518)

[ghstack-poisoned]

iseeyuan reviewed (approved):
LGTM

Commit:
Update on "[Operator Versioning][Edge] Implement register function for upgrader"




This pr implement the register function for upgrader so it can be used at loading stage

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518/)

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Implement register function for upgrader"




This pr implement the register function for upgrader so it can be used at loading stage

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518/)

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Implement register function for upgrader"




This pr implement the register function for upgrader so it can be used at loading stage

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518/)

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Implement register function for upgrader"




This pr implement the register function for upgrader so it can be used at loading stage

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518/)

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Implement register function for upgrader"




This pr implement the register function for upgrader so it can be used at loading stage

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518/)

Differential Revision: [D32092518](https://our.internmc.facebook.com/intern/diff/D32092518)

[ghstack-poisoned]
